### PR TITLE
feat: Add JSON configuration parser

### PR DESCRIPTION
**Pull request by FumingPower3925**
Adds a new parser to support loading configuration values from JSON files.

Includes:
- `json` package with the parser implementation.
- Comprehensive unit tests covering various JSON structures and edge cases.
- Fix for handling `null` values in the core `ToString` conversion function.

### CODE REVIEW

File: json/parse.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	p.conv = conv
+	p.awaited = keys
+
+	return p.parse(data)
+}
+
+func (p *Parser) parse(data []byte) (found, unknown map[string]string, err error) {
+	var settings any
+
+	tempFile, err := os.CreateTemp("", "json-input-*.json")
```
Comment (chaindead) [inline json/parse.go:56]: what is the purpose to create temp file?

File: conv.go [hunk 2]
```diff
@@ -16,6 +16,9 @@ import (
 //
 // ToString is used internally by zerocfg for representing option values as strings, including when passing ToString to custom parsers and for rendering configuration output.
 func ToString(v any) string {
+	if v == nil {
```
Comment (chaindead) [inline conv.go:19]: do not pass nil to this func

File: json/parse_test.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
+package json_test
```
Comment (chaindead) [inline json/parse_test.go:1]: simplify testcases (take yaml package as reference)

- we do not test encoding/json
- delete all tests except base that do not expand coverage, add test for corner cases like null (1 or 2)

File: conv.go (hunk 2 above)
Comment (chaindead) [reply conv.go:19]: skip such values

**Comment by chaindead** [comment]
also, why you need raw json parser? 

i understand if it is remote source or smth like this that provides json configuration, but i do not see cases of using file in json over yaml

**Comment by FumingPower3925** [comment]
> also, why you need raw json parser?
> 
> i understand if it is remote source or smth like this that provides json configuration, but i do not see cases of using file in json over yaml

Well, in my workplace we use JSON for configs as a standard, so I am kind of used to it. I do not have anything against YAML, however, I find JSON configs to handle better really complex application configs, I find them easier to maintain.

File: json/parse.go (hunk 1 above)
Comment (FumingPower3925) [reply json/parse.go:56]: I intiially did that because I was having problems with the parsing of numbers from the json (1e6 vs. 1e+06). However, after your comment I realized that this can be done by using bytes.NewReader(data) directly with the json.Decoder. I am going to correct this

File: json/parse.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	return found, unknown, nil
+}
+
+func (p *Parser) flatten(settings map[string]any) (found, unknown map[string]string) {
+	found, unknown = make(map[string]string), make(map[string]string)
+	p.flattenDFS(settings, "", found, unknown)
+	return found, unknown
+}
+
+func (p *Parser) flattenDFS(m map[string]any, prefix string, found, unknown map[string]string) {
```
Comment (chaindead) [inline json/parse.go:79]: check how yaml is provider is made, also change plz parser to provider, 

try to avoid using else, use do not need so complex code

Comment (chaindead) [reply json/parse.go:79]: i think here u can copy paste - yaml flattenDFS

File: json/parse.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	p.awaited = keys
+
+	return p.parse(data)
+}
+
+func (p *Parser) parse(data []byte) (found, unknown map[string]string, err error) {
+	var settings any
+
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.UseNumber()
```
Comment (chaindead) [inline json/parse.go:58]: does this works well with zfg.ToString func?

File: json/parse.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	if err = decoder.Decode(&settings); err != nil {
```
Comment (chaindead) [inline json/parse.go:59]: why not just json.Unmarshal?

File: json/parse.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		if err.Error() == "EOF" && len(data) == 0 {
```
Comment (chaindead) [inline json/parse.go:60]: errors.Is(...) and why eof is ok?

File: json/parse_test.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	assert.Empty(t, found)
+	assert.Empty(t, unknown)
+}
+
+func TestParse_FileNotExist(t *testing.T) {
+	nonExistentPath := "no_such_file.json"
+	p := json.New(&nonExistentPath)
+
+	found, unknown, err := p.Parse(map[string]bool{"some.key": true}, zfg.ToString)
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (FumingPower3925) [reply json/parse_test.go:144]: In the code I handled the non existence of the path and other errors when opening the file as a "none" event instead of an error. I will change it to better align with your implementation

File: json/parse.go (hunk 5 above)
Comment (FumingPower3925) [reply json/parse.go:58]: The interaction is fine: json.Number is a string, and ToString handles strings. The benefit is fidelity to the original number string. I will explain it better in the unmarshal comment.

File: json/parse.go (hunk 6 above)
Comment (FumingPower3925) [reply json/parse.go:59]: Using Umarshal (as far as I know) would convert all number types and treat them as float64. This loses part of the expresivness that Json number have, but if you prefer it that way I can change it to use Unmarshal. I would simplify the code.

File: json/parse.go (hunk 7 above)
Comment (FumingPower3925) [reply json/parse.go:60]: I did not treat empty jsons as errors, but I will in order to align better with the Yaml implementation.

**Comment by FumingPower3925** [comment]
I've added the changes you requested on treating as errors missing file, empty file and malformed file.

**I've not** changed from decode to unmarshall as I still believe that maintaining the expressiveness of the number values in json is important. However, as I said in another comment, if you tell me so, I will change it to use Unmarshall.

**Comment by chaindead** [comment]
i thought how json support should be implemented, here what i think

example code to fetch from url with file support
```go
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
)

func fetchJSON(u *url.URL) ([]byte, error) {
	var reader io.ReadCloser
	var err error

	switch u.Scheme {
	case "http", "https":
		resp, err := http.Get(u.String())
		if err != nil {
			return nil, fmt.Errorf("http get failed: %w", err)
		}
		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			return nil, fmt.Errorf("http status: %s", resp.Status)
		}
		reader = resp.Body
	case "file":
		// Normalize the file path
		path := u.Path
		if u.Host != "" {
			// Handle Windows UNC-style paths like file://host/path
			path = "//" + u.Host + u.Path
		}
		reader, err = os.Open(path)
		if err != nil {
			return nil, fmt.Errorf("file open failed: %w", err)
		}
	default:
		return nil, fmt.Errorf("unsupported URL scheme: %s", u.Scheme)
	}
	defer reader.Close()

	data, err := io.ReadAll(reader)
	if err != nil {
		return nil, fmt.Errorf("read failed: %w", err)
	}
	return data, nil
}
```

- there needed to add URL configuration option support like

url:=zfg.URL("name", "file:///a/b/c.json", "path to json config")
and json provider api like this
json.New(url)

- needed to add json provider description to readme section

**Comment by FumingPower3925** [comment]
I've added the URL support as you described. I find it a great idea!

**Comment by chaindead** [comment]
add test to z_test.go for url, use struct like 

```go
type urlVale url.Url
```

**Comment by FumingPower3925** [comment]
I've modified z_url to use `type urlVale url.Url`.
Added the tests in z_test.go for url.
